Sources/TIMEncryptedStorage/TIM/KeyService/TIMKeyService.swift

```diff
@@ -28,52 +28,77 @@ public final class TIMKeyService : TIMKeyServiceProtocol {
 
     /// The `URLSession` to perform the network requests to the Trifork Identity Manager KeyService.
     private let urlSession: URLSession
+    /// The chosen URLSessionConfiguration chosen in the initialiser, used for re-initialising the URLSession if needed.
+    private let urlSessionConfiguration: URLSessionConfiguration
+    /// The number of times a failed url request is reattempted before an error is thrown
+    private let requestRetryAttempts: Int
 
     /// Sets the configuration of the key service. This should be called before you call any other functions on this class.
     /// - Parameter configuration: The configuration.
     /// - Parameter networkSession: The `URLSession` to perform the network requests to the Trifork Identity Manager KeyService. Default is `URLSession(configuration: .ephemeral)`
-    public init(configuration: TIMKeyServiceConfiguration, urlSession: URLSession = URLSession(configuration: .ephemeral)) {
+    public init(
+        configuration: TIMKeyServiceConfiguration,
+        urlSession: URLSession = URLSession(configuration: .ephemeral),
+        requestRetryAttempts: Int = 1
+    ) {
         self.configuration = configuration
         self.urlSession = urlSession
+        self.urlSessionConfiguration = urlSession.configuration
+        self.requestRetryAttempts = requestRetryAttempts
 
         guard verifyConfiguration(configuration) else {
             fatalError("TIMKeyService configuration is invalid!")
         }
     }
 
-    private func request<T: Decodable>(_ url: URL, parameters: [String : String], completion: @escaping (Result<T, TIMKeyServiceError>) -> Void) {
-        var request = URLRequest(url: url)
-        request.httpMethod = "POST"
-        request.addValue("application/json", forHTTPHeaderField: "Content-Type")
-        request.httpBody = try? JSONEncoder().encode(parameters)
-
-        let task = urlSession.dataTask(with: request) { (data, response, error) in
-            guard let response = response as? HTTPURLResponse else {
-                DispatchQueue.main.async {
-                    completion(.failure(mapKeyServiceError(error)))
+    private func request<T: Decodable>(_ url: URL, parameters: [String: String], completion: @escaping (Result<T, TIMKeyServiceError>) -> Void) {
+        // Helper method to perform the request and handle retries
+        func performRequest(with urlSession: URLSession, retryCount: Int) {
+            var request = URLRequest(url: url)
+            request.httpMethod = "POST"
+            request.addValue("application/json", forHTTPHeaderField: "Content-Type")
+            request.httpBody = try? JSONEncoder().encode(parameters)
```

Comment:
While at it, should the encode throw an exception and the method return a TIMKeyServerError?

Reply:
I'll make this a separate PR :)